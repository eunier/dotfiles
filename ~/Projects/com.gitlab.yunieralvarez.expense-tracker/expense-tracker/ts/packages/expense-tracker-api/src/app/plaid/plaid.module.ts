import { Module } from '@nestjs/common';
import { ConfigModule } from '../config/config.module';
import { PlaidApiClassProvider } from './providers/plaid-api-class.provider';
import { PlaidService } from './services/plaid/plaid.service';

@Module({
	imports: [ConfigModule],
	providers: [PlaidService, PlaidApiClassProvider],
	exports: [PlaidService],
})
export class PlaidModule {}
