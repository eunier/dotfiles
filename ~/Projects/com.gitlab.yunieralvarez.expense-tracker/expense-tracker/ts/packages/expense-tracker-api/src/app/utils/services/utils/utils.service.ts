import * as util from 'node:util';
import { Injectable } from '@nestjs/common';

@Injectable()
export class Utils {
	public delay(ms: number): Promise<void> {
		return new Promise((resolve) => setTimeout(resolve, ms));
	}

	public static *range(length: number) {
		for (let i = 0; i < length; i++) {
			yield i;
		}
	}

	public static inspectObject(
		...args: Parameters<(typeof util)['inspect']>
	): string {
		return util.inspect(...args);
	}

	public static genSeedFromString(string: string) {
		let hash = 0;

		for (let i = 0; i < string.length; i++) {
			const char = string.charCodeAt(i);
			hash = (hash << 5) - hash + char;
			hash = hash & hash; // Convert to 32bit integer
		}

		return hash;
	}
}
