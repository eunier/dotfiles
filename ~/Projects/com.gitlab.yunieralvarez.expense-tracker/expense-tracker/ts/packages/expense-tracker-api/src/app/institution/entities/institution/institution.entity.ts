import { Entity, EntityRepositoryType, Property } from '@mikro-orm/core';
import { InstitutionRepository } from '~/app/institution/repositories/institution/institution.repository';
import { BaseEntity } from '~/app/orm/entities/base/base.entity';
import { PlaidLib } from '~/app/plaid/libs/plaid.lib';

@Entity({
	tableName: InstitutionEntity.metadata.tableName,
	repository: InstitutionEntity.metadata.repositoryFn,
})
export class InstitutionEntity extends BaseEntity {
	public static metadata = {
		repository: InstitutionRepository,
		repositoryFn: () => InstitutionEntity.metadata.repository,
		tableName: 'institution',
	};

	[EntityRepositoryType]? = InstitutionEntity.metadata.repository;

	@Property()
	public plaidId: string;

	@Property()
	public name: string;

	public constructor(plaidId: string, name: string) {
		super();
		this.plaidId = plaidId;
		this.name = name;
	}

	public matchesPlaidInstitution(
		plaidInstitution: PlaidLib.InstitutionsGetByIdResponse,
	) {
		return (
			this.name === plaidInstitution.institution.name &&
			this.plaidId === plaidInstitution.institution.institution_id
		);
	}

	public static fromPlaidGetInstitutionByIdResponse(
		response: PlaidLib.InstitutionsGetByIdResponse,
	) {
		return new InstitutionEntity(
			response.institution.institution_id,
			response.institution.name,
		);
	}
}
