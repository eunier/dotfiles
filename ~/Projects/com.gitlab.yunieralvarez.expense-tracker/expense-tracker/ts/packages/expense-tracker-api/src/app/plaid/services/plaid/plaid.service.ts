import { Inject, Injectable, Logger } from '@nestjs/common';
import { AxiosResponse } from 'axios';
import { Config } from '~/app/config/services/config/config.service';
import { PlaidLib } from '~/app/plaid/libs/plaid.lib';
import { ExchangePublicTokenResponse } from '~/app/plaid/models/exchange-public-token-response/exchange-public-token-response.model';
import { TransactionFetchResponse } from '~/app/plaid/models/fetch-transactions-response/transaction-fetch-response.model';
import { PLAID_API_CLASS_TOKEN } from '~/app/plaid/providers/plaid-api-class.provider';

@Injectable()
export class PlaidService {
	private readonly logger = new Logger(PlaidService.name);
	private readonly configuration: PlaidLib.Configuration;
	private readonly plaidClient: PlaidLib.PlaidApi;

	public constructor(
		@Inject(PLAID_API_CLASS_TOKEN)
		private readonly plaidApiClass: typeof PlaidLib.PlaidApi,
		private readonly config: Config,
	) {
		this.configuration = new PlaidLib.Configuration({
			basePath: PlaidLib.PlaidEnvironments.sandbox,
			baseOptions: {
				headers: {
					'PLAID-CLIENT-ID': this.config.plaid.clientId,
					'PLAID-SECRET': this.config.plaid.secret,
					'Plaid-Version': this.config.plaid.version,
				},
			},
		});

		this.plaidClient = new this.plaidApiClass(this.configuration);
	}

	/**
	 * Crate a Plaid link token.
	 * @param userId The user id.
	 * @returns The Plaid link token.
	 */
	public async createLinkToken(userId: string) {
		const request: PlaidLib.LinkTokenCreateRequest = {
			client_name: this.config.app.plaidClientName,
			country_codes: this.config.plaid.countryCodes,
			language: 'en',
			products: this.config.plaid.products,
			user: { client_user_id: userId },
		};

		const linkTokenResponse: AxiosResponse<
			PlaidLib.LinkTokenCreateResponse,
			unknown
		> = await this.plaidClient.linkTokenCreate(request);

		return linkTokenResponse.data;
	}

	/**
	 * Exchange a Plaid public token for a Plaid access token.
	 * @param publicToken Plaid public token.
	 * @returns Plaid access token.
	 */
	public async exchangePublicToken(publicToken: string) {
		const request: PlaidLib.ItemPublicTokenExchangeRequest = {
			public_token: publicToken,
		};

		const plaidResponse =
			await this.plaidClient.itemPublicTokenExchange(request);

		const response =
			ExchangePublicTokenResponse.fromPlaidItemPublicTokenExchangeResponse(
				plaidResponse.data,
			);

		return response;
	}

	/**
	 * Get an institution by id.
	 * @param institutionId The institution id.
	 * @returns The institution
	 */
	public async getInstitutionById(institutionId: string) {
		const req: PlaidLib.InstitutionsGetByIdRequest = {
			institution_id: institutionId,
			country_codes: this.config.plaid.countryCodes,
		};

		const res = await this.plaidClient.institutionsGetById(req);
		return res.data;
	}

	/**
	 * Fetch transactions.
	 * @param sourceAccessToken The access token for the source.
	 * @param cursor The next transaction cursor.
	 * @returns Fetched transactions.
	 */
	public async fetchTransactions(sourceAccessToken: string, cursor = '') {
		let added: PlaidLib.Transaction[] = [];
		let modified: PlaidLib.Transaction[] = [];
		let removed: PlaidLib.RemovedTransaction[] = [];
		let nextCursor = cursor;
		let hasMore = false;
		const count = this.config.app.plaidTransactionSyncCount;

		do {
			const res = await this.plaidClient.transactionsSync({
				access_token: sourceAccessToken,
				count,
				cursor: nextCursor,
				options: { include_original_description: true },
			});

			added = added.concat(res.data.added);
			modified = modified.concat(res.data.modified);
			removed = removed.concat(res.data.removed);
			nextCursor = res.data.next_cursor;
			hasMore = res.data.has_more;

			this.logger.log(
				`Received ${res.data.added.length} added, ${res.data.modified.length} modified, and ${res.data.removed.length} removed transaction(s)`,
			);

			this.logger.log(
				`Transaction sync response ${res.data.has_more ? 'had more' : 'does not have more'} transaction(s)`,
			);
		} while (hasMore);

		this.logger.log(
			`Received a total of ${added.length} added, ${modified.length} modified, and ${removed.length} removed transaction(s)`,
		);

		return new TransactionFetchResponse(added, modified, removed, nextCursor);
	}
}
