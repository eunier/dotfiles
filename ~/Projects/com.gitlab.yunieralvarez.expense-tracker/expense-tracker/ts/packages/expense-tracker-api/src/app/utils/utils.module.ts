import { Module } from '@nestjs/common';
import { Utils } from '~/app/utils/services/utils/utils.service';

@Module({
	providers: [Utils],
	exports: [Utils],
})
export class UtilsModule {}
