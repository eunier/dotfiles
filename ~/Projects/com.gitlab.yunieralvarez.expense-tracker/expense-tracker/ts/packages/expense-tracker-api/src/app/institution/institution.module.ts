import { MikroOrmModule } from '@mikro-orm/nestjs';
import { Module } from '@nestjs/common';
import { InstitutionEntity } from '~/app/institution/entities/institution/institution.entity';
import { InstitutionRepositoryProvider } from '~/app/institution/providers/institution-repository.provider';
import { InstitutionService } from '~/app/institution/services/institution/institution.service';
import { PlaidModule } from '~/app/plaid/plaid.module';

@Module({
	imports: [PlaidModule, MikroOrmModule.forFeature([InstitutionEntity])],
	providers: [InstitutionService, InstitutionRepositoryProvider],
	exports: [InstitutionService],
})
export class InstitutionModule {}
