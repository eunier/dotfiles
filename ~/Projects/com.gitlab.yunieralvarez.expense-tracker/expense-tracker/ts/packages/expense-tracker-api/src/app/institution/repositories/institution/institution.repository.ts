import { EntityRepository } from '@mikro-orm/postgresql';
import { Injectable } from '@nestjs/common';
import { InstitutionEntity } from '../../entities/institution/institution.entity';

@Injectable()
export class InstitutionRepository extends EntityRepository<InstitutionEntity> {}
