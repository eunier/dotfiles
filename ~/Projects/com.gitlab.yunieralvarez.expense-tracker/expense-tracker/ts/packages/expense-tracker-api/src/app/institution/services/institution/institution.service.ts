import { EntityManager } from '@mikro-orm/postgresql';
import { Injectable, Logger } from '@nestjs/common';
import { InstitutionEntity } from '~/app/institution/entities/institution/institution.entity';
import { InstitutionRepository } from '~/app/institution/repositories/institution/institution.repository';
import { PlaidService } from '~/app/plaid/services/plaid/plaid.service';

@Injectable()
export class InstitutionService {
	private logger = new Logger(InstitutionService.name);

	constructor(
		private readonly entityManager: EntityManager,
		private readonly institutionRepository: InstitutionRepository,
		private readonly plaidService: PlaidService,
	) {}

	/**
	 * Sync plaid institution info with the app's.
	 *
	 * 1. Get institution info from plaid.
	 * 2. If not in the app database, save it, else, do nothing.
	 * @param plaidId The plaid institution id.
	 */
	public async getAndUpsertInstitution(plaidId: string) {
		const response = await this.plaidService.getInstitutionById(plaidId);
		this.logger.verbose(`Fetched institution: ${response.institution.name}`);

		let institution = await this.institutionRepository.findOne({
			plaidId,
		});

		if (!institution) {
			this.logger.verbose(`${response.institution.name} is a new institution`);
			this.logger.verbose(`Saving ${response.institution.name} institution...`);
		} else if (!institution.matchesPlaidInstitution(response)) {
			this.logger.verbose(
				`Institution ${institution.name} plaidId does not match plaid institution`,
			);

			this.logger.verbose(`Updating ${institution.name} institution...`);
		}

		institution =
			InstitutionEntity.fromPlaidGetInstitutionByIdResponse(response);

		await this.entityManager.persistAndFlush(institution);
		return institution;
	}
}
