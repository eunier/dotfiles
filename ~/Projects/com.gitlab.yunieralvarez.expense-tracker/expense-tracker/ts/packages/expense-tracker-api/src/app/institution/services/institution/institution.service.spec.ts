import { faker } from '@faker-js/faker/.';
import { EntityManager } from '@mikro-orm/postgresql';
import { Logger } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { DeepMockProxy, MockProxy, mock, mockDeep } from 'jest-mock-extended';
import { InstitutionEntity } from '~/app/institution/entities/institution/institution.entity';
import { InstitutionRepository } from '~/app/institution/repositories/institution/institution.repository';
import { InstitutionService } from '~/app/institution/services/institution/institution.service';
import { PlaidLib } from '~/app/plaid/libs/plaid.lib';
import { PlaidService } from '~/app/plaid/services/plaid/plaid.service';
import { Utils } from '~/app/utils/services/utils/utils.service';

beforeEach(() => {
	jest.useFakeTimers();

	jest.setSystemTime(
		new Date(faker.number.int({ min: 1970, max: 2000 }), 1, 1).getTime(),
	);

	faker.seed(
		Utils.genSeedFromString(
			expect.getState().currentTestName ?? InstitutionService.name,
		),
	);
});

describe('InstitutionService', () => {
	let institutionService: InstitutionService;
	let logger: DeepMockProxy<typeof Logger>;
	let plaidService: MockProxy<PlaidService>;
	let institutionRepository: MockProxy<InstitutionRepository>;
	let entityManager: MockProxy<EntityManager>;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				InstitutionService,
				{
					provide: EntityManager,
					useValue: mock(EntityManager),
				},
				{
					provide: InstitutionRepository,
					useValue: mock(InstitutionRepository),
				},
				{
					provide: PlaidService,
					useValue: mock(PlaidService),
				},
			],
		}).compile();

		institutionService = module.get<InstitutionService>(InstitutionService);
		logger = mockDeep(Logger);
		institutionService['logger'] = logger as unknown as Logger;

		plaidService = module.get<PlaidService>(
			PlaidService,
		) as MockProxy<PlaidService>;

		institutionRepository = module.get<InstitutionRepository>(
			InstitutionRepository,
		) as MockProxy<InstitutionRepository>;

		entityManager = module.get<EntityManager>(
			EntityManager,
		) as MockProxy<EntityManager>;
	});

	describe('getAndUpsertInstitution', () => {
		describe('when institution is not in the database', () => {
			it('should log', async () => {
				const plaidId = faker.string.nanoid();

				{
					const resInstitution = mockDeep<PlaidLib.Institution>();
					resInstitution.institution_id = plaidId;
					resInstitution.name = faker.company.name();

					const res = mockDeep<PlaidLib.InstitutionsGetByIdResponse>();
					res.institution = resInstitution;
					plaidService.getInstitutionById.mockResolvedValue(res);
				}

				institutionRepository.findOne.mockResolvedValue(null);
				const loggerSpy = jest.spyOn(logger, 'verbose');
				await institutionService.getAndUpsertInstitution(plaidId);
				const loggerArgs = loggerSpy.mock.calls;
				expect(loggerArgs).toMatchSnapshot();
			});
		});

		describe('when institution is already in the database', () => {
			describe('when institution nodes not match', () => {
				it('should log', async () => {
					const plaidId = faker.string.nanoid();

					{
						const resInstitution = mockDeep<PlaidLib.Institution>();
						resInstitution.institution_id = plaidId;
						resInstitution.name = faker.company.name();

						const res = mockDeep<PlaidLib.InstitutionsGetByIdResponse>();
						res.institution = resInstitution;
						plaidService.getInstitutionById.mockResolvedValue(res);
					}

					{
						const institution = mock<InstitutionEntity>({
							matchesPlaidInstitution:
								InstitutionEntity.prototype.matchesPlaidInstitution,
						});

						institution.id = faker.string.uuid();
						institution.name = faker.company.name();
						institution.plaidId = plaidId;
						institutionRepository.findOne.mockResolvedValue(institution);
					}

					const loggerSpy = jest.spyOn(logger, 'verbose');
					await institutionService.getAndUpsertInstitution(plaidId);
					const loggerArgs = loggerSpy.mock.calls;
					expect(loggerArgs).toMatchSnapshot();
				});
			});
		});
	});
});

describe('InstitutionService (legacy)', () => {
	let institutionService: InstitutionService;
	let plaidService: PlaidService;
	let entityManager: EntityManager;
	let institutionRepository: InstitutionRepository;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				InstitutionService,
				{
					provide: EntityManager,
					useValue: {
						persistAndFlush: jest.fn(),
					} satisfies Partial<EntityManager>,
				},
				{
					provide: InstitutionRepository,
					useValue: {
						findOne: jest.fn(),
					} satisfies Partial<InstitutionRepository>,
				},
				{
					provide: PlaidService,
					useValue: {
						getInstitutionById: jest.fn(),
					} satisfies Partial<PlaidService>,
				},
			],
		}).compile();

		institutionService = module.get<InstitutionService>(InstitutionService);
		plaidService = module.get<PlaidService>(PlaidService);
		entityManager = module.get<EntityManager>(EntityManager);

		institutionRepository = module.get<InstitutionRepository>(
			InstitutionRepository,
		);
	});

	it('should be defined', () => {
		expect(institutionService).toBeDefined();
	});
});
