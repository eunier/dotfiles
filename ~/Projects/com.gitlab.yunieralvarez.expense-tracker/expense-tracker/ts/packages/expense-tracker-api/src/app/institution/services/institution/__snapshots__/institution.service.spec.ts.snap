// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InstitutionService getAndUpsertInstitution when institution is already in the database when institution nodes not match should log 1`] = `
[
  [
    "Fetched institution: Roberts LLC",
  ],
  [
    "Roberts LLC is a new institution",
  ],
  [
    "Saving Roberts LLC institution...",
  ],
  [
    "Fetched institution: Langosh - Jacobi",
  ],
  [
    "Institution Mosciski, Langosh and Gislason plaidId does not match plaid institution",
  ],
  [
    "Updating Mosciski, Langosh and Gislason institution...",
  ],
]
`;

exports[`InstitutionService getAndUpsertInstitution when institution is not in the database should log 1`] = `
[
  [
    "Fetched institution: Roberts LLC",
  ],
  [
    "Roberts LLC is a new institution",
  ],
  [
    "Saving Roberts LLC institution...",
  ],
]
`;
